Mcmc object (#17)
rowlandseymour/speedyBBT: vignettes/getting-started.Rmd

---

> title: "Getting started with `speedyBBT`"

Reviewer: I'm not keen on using code markup for a package name as there are only certain cases where we can use the bare name (e.g. `library()` calls). I suggest to use boldface instead, but no need to mark up here as the title will be bold in any case.

output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{my-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(speedyBBT)
library(coda)
```

> `speedyBBT` is a package for Bayesian Bradley-Terry modeling. It provides functions for fitting the Bradley-Terry model using Markov Chain Monte Carlo (MCMC) methods, allowing for inference on player abilities and hyperparameters. It uses latent Polya-Gamma variables to facilitate efficient sampling.

Reviewer: Maybe we should use English spelling given the lead author? I resume English spelling is used in the JSS article, but haven't checked.

```suggestion
`speedyBBT` is a package for Bayesian Bradley-Terry modelling. It provides functions for fitting the Bradley-Terry model using Markov Chain Monte Carlo (MCMC) methods, allowing for inference on player abilities and hyperparameters. It uses latent Polya-Gamma variables to facilitate efficient sampling.
```

The `speedyBBTm` function can be used to fit a Bradley-Terry model to the data. You can specify the outcome variable, the items involved in each comparison, and any prior information you have about the items' rating. The function will return an MCMC object containing samples from the posterior distribution of the item qualities. For this example, you first need to construct an adjacency matrix and then use it to create a covariance matrix for the prior distribution of the item qualities.

```{r covariance}
# Construct covariance matrix
expA <- expm::expm(forcedMarriage$adjacencyMatrix)
sigma <- diag(diag(expA)^-0.5) %*% expA %*% diag(diag(expA)^-0.5)
```

> Next, you can fit the model using the `speedyBBTm` function. You will need to specify the outcome variable, the items involved in each comparison, and the prior covariance matrix for the item qualities. You can also specify the number of MCMC iterations to run. This example uses the default values for the hyperparameters of the prior distribution, but you can also specify your own values if you have prior knowledge about the items' rating using the `hyperparameter` argument.

Reviewer: Add parentheses to function calls for clarity

```suggestion
Next, you can fit the model using the `speedyBBTm()` function. You will need to specify the outcome variable, the items involved in each comparison, and the prior covariance matrix for the item qualities. You can also specify the number of MCMC iterations to run. This example uses the default values for the hyperparameters of the prior distribution, but you can also specify your own values if you have prior knowledge about the items' rating using the `hyperparameter` argument.
```

```{r fit_model}
# Fit model
forcedMarriageModel <- speedyBBTm(
  outcome = rep(1, length(forcedMarriage$comparisons$win)),
  player1 = forcedMarriage$comparisons$win,
  player2 = forcedMarriage$comparisons$lost,
  player.prior.var = sigma,
  n.iter = 10000,
  burn.in = 100
)
```

To view some preliminary information about the results, you can calculate the means of the item quality parameters from the MCMC samples and summarise them. This will give you an idea of the estimated qualities of the items based on the comparisons in the dataset.

```{r view_output}
forcedMarriageModelMeans <- colMeans(forcedMarriageModel[, grep("lambda", varnames(forcedMarriageModel))])
forcedMarriageModelMeans
summary(forcedMarriageModel)
```

> You can run some MCMC diagnostics to check convergence. The `speedyBBTm` function returns an MCMC object that can be used with the `coda` package for diagnostics. You can use the `plot()` function to visualize the trace plots and density plots of the MCMC samples. 

Reviewer: ```suggestion
You can run some MCMC diagnostics to check convergence. The `speedyBBTm()` function returns an MCMC object that can be used with the `coda` package for diagnostics. You can use the `plot()` function to visualize the trace plots and density plots of the MCMC samples. 
```

```{r mcmc_diagnostics, fig.width=8, fig.height=6}
# Trace plots

> plot(forcedMarriageModel[, paste("lambda[", c(10, 20, 30, 40), "]", sep = "")], main = "Trace plots for item quality parameters")

Reviewer: ```suggestion
plot(forcedMarriageModel[, paste0("lambda[", c(10, 20, 30, 40), "]")], main = "Trace plots for item quality parameters")
```

```

> You can use the `effectiveSize()` function to calculate the estimated effective sample size for convergence. Another example diagnostic illustrated here is the Geweke diagnostic, calculated using the `gewecke.diag()` function from `coda`. This diagnostic compares the different between the two sample means in the first 10% and the last 50% of iterations in the Markov chain to asses convergence. For more information about available plots and diagnostics, you can refer to the [`coda` ](https://github.com/cran/coda) package [documentation](https://cran.r-project.org/web/packages/coda/index.html).

Reviewer: ```suggestion
You can use the `effectiveSize()` function to calculate the estimated effective sample size for convergence. Another example diagnostic illustrated here is the Geweke diagnostic, calculated using the `gewecke.diag()` function from `coda`. This diagnostic compares the difference between the two sample means in the first 10% and the last 50% of iterations in the Markov chain to assess convergence. For more information about available plots and diagnostics, you can refer to the [**coda** package documentation](https://cran.r-project.org/package=coda).
```

```{r some_diags}
effectiveSize(forcedMarriageModel)
geweke.diag(forcedMarriageModel)
``` 

> You can get additional, colourful plots by installing `ggplot2` and `bayesplot` and calling the functions from `bayesplot` on the `mcmc` object returned by `speedyBBTm`. For example, you can use the `mcmc_trace()` function to create trace plots and the `mcmc_dens_overlay()` function to create density plots of the MCMC samples. For more information, see the `bayesplot` package [documentation](https://mc-stan.org/bayesplot/).

Reviewer: Avoid using the same word for hyperlinks as that's unhelpful for screenreaders

```suggestion
You can get additional, colourful plots by installing `ggplot2` and `bayesplot` and calling the functions from `bayesplot` on the `mcmc` object returned by `speedyBBTm`. For example, you can use the `mcmc_trace()` function to create trace plots and the `mcmc_dens_overlay()` function to create density plots of the MCMC samples. For more information, see the [**bayesplot** package documentation](https://mc-stan.org/bayesplot/).
```
